### PR TITLE
unix,win: use static_assert when available

### PR DESCRIPTION
**Pull request by bnoordhuis**
Fixes: https://github.com/libuv/libuv/issues/3131

### CODE REVIEW

**Comment by alzix** [comment]
please note `static_assert(...)` requires an include to `assert.h` when building with `Apple clang version 12.0.0 (clang-1200.0.32.29)`.
This is what i found during my #3186 testing.

**Comment by bnoordhuis** [comment]
uv-common.h includes assert.h so that should be no problem. I added a fix for android because it seems to be missing static_assert.

**Review by cjihrig — APPROVED** [review]

**Comment by bnoordhuis** [comment]
> this is where configure time detection comes handy

I'd have to do it in two places though: CMakeLists.txt and autoconf.ac.

**Review by vtjnash — APPROVED** [review]
CI: https://ci.nodejs.org/job/libuv-test-commit/2158/

**Review by vtjnash — CHANGES_REQUESTED** [review]
This seems to have fared pretty poorly in CI

File: src/uv-common.h [hunk 1]
```diff
@@ -57,8 +57,13 @@ extern int snprintf(char*, size_t, const char*, ...);
 #define container_of(ptr, type, member) \
   ((type *) ((char *) (ptr) - offsetof(type, member)))
 
+/* Android NDK 20 doesn't seem to know static_assert. */
+#if defined(__GNUC__) && !defined(__ANDROID__)
+#define STATIC_ASSERT(expr) static_assert(expr, #expr)
```
Comment (vtjnash) [inline src/uv-common.h:62]:
```suggestion
/* C11 defines static_assert to be a macro which calls _Static_assert. */
#if defined(static_assert)
#define STATIC_ASSERT(expr) static_assert(expr, #expr)
```

**Review by vtjnash — APPROVED** [review]

**Pull request merged**